### PR TITLE
UPSTREAM: <carry>: openshift: Record events for actuator CUD operations

### PR DESCRIPTION
**Pull request by ingvagabund**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by ingvagabund** [comment]
`ci/prow/e2e-azure` is constantly failing on storage tests (`PV` or `Inline-volume). Though, there is always 6 nodes deploy and machine controller logs does not show anything suspicious.

File: pkg/cloud/azure/actuators/machine/actuator.go [hunk 1]
```diff
@@ -40,23 +51,42 @@ import (
...
+		eventRecorder:     params.EventRecorder,
+		reconcilerBuilder: params.ReconcilerBuilder,
+	}
+}
+
+// Set corresponding event based on error. It also returns the original error
+// for convenience, so callers can do "return handleMachineError(...)".
+func (a *Actuator) handleMachineError(machine *machinev1.Machine, err *apierrors.MachineError, eventAction string) error {
+	if eventAction != noEventAction {
+		fmt.Printf("fuuu: %v\n", err.Reason)
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine/actuator.go]: Debug?

File: pkg/cloud/azure/actuators/machine/actuator.go [hunk 2]
```diff
@@ -125,19 +159,21 @@ func (a *Actuator) Update(ctx context.Context, cluster *clusterv1.Cluster, machi
 		CoreClient: a.coreClient,
 	})
 	if err != nil {
-		return errors.Errorf("failed to create scope: %+v", err)
+		return a.handleMachineError(machine, apierrors.UpdateMachine("failed to create scope: %v", err), updateEventAction)
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine/actuator.go]: Message should be `failed to update`?

File: pkg/cloud/azure/actuators/machine/actuator_test.go [hunk 3]
```diff
@@ -33,12 +35,17 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
 	"sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1alpha1"
 	providerv1 "sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1alpha1"
 	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure"
 	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/actuators"
 	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure/services/virtualmachines"
 	controllerfake "sigs.k8s.io/controller-runtime/pkg/client/fake"
+
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine/actuator_test.go]: nit: drop blank line.

File: pkg/cloud/azure/actuators/machine/actuator_test.go [hunk 4]
```diff
@@ -578,6 +585,187 @@ func TestCustomDataFailures(t *testing.T) {
 	}
 }
 
+func TestMachineEvents(t *testing.T) {
+
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine/actuator_test.go]: drop blank line.

File: pkg/cloud/azure/actuators/machine/stubs.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+- path: /root/node_bootstrap/node_settings.yaml
+  owner: 'root:root'
+  permissions: '0640'
+  content: |
+    node_config_name: node-config-master
+runcmd:
+- [ cat, /root/node_bootstrap/node_settings.yaml]
+`
+
+// IntPtr returns a pointer to an int
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine/stubs.go]: Does this need to be exported?

File: pkg/cloud/azure/actuators/machine/actuator.go (hunk 2 above)
Comment (ingvagabund) [reply pkg/cloud/azure/actuators/machine/actuator.go]: Machine scope is always created, never updated. So, it's correct to use `failed to create scope`.

File: pkg/cloud/azure/actuators/machine/actuator.go [hunk 6]
```diff
@@ -70,18 +98,20 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
 		CoreClient: a.coreClient,
 	})
 	if err != nil {
-		return errors.Errorf("failed to create scope: %+v", err)
+		return a.handleMachineError(machine, apierrors.CreateMachine("failed to create scope: %v", err), createEventAction)
```
Comment (enxebre) [inline pkg/cloud/azure/actuators/machine/actuator.go]: it'd be nice to include the `machine.Name()` in this message

File: pkg/cloud/azure/actuators/machine/actuator.go [hunk 7]
```diff
@@ -70,18 +98,20 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
...
 	}
 	defer scope.Close()
 
-	err = NewReconciler(scope).Create(context.Background())
+	err = a.reconcilerBuilder(scope).Create(context.Background())
 	if err != nil {
-		klog.Errorf("failed to reconcile machine %s: %v", machine.Name, err)
+		a.handleMachineError(machine, apierrors.CreateMachine("failed to reconcile machine %s: %v", machine.Name, err), createEventAction)
```
Comment (enxebre) [inline pkg/cloud/azure/actuators/machine/actuator.go]: very nit: personally I prefer to use the method to retrieve the name `machine.Name()` and `%q` for names

File: pkg/cloud/azure/actuators/machine/actuator.go [hunk 8]
```diff
@@ -70,18 +98,20 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
...
 		return &controllerError.RequeueAfterError{
 			RequeueAfter: time.Minute,
 		}
 	}
 
+	a.eventRecorder.Eventf(machine, corev1.EventTypeNormal, "Created", "Created machine %v", machine.Name)
```
Comment (enxebre) [inline pkg/cloud/azure/actuators/machine/actuator.go]: nitL: can we use either %s, %v or %q consitently for priting names? my preference is %q

File: pkg/cloud/azure/actuators/machine/actuator.go [hunk 9]
```diff
@@ -96,19 +126,21 @@ func (a *Actuator) Delete(ctx context.Context, cluster *clusterv1.Cluster, machi
 		CoreClient: a.coreClient,
 	})
 	if err != nil {
-		return errors.Wrapf(err, "failed to create scope")
+		return a.handleMachineError(machine, apierrors.DeleteMachine("failed to create scope: %v", err), deleteEventAction)
```
Comment (enxebre) [inline pkg/cloud/azure/actuators/machine/actuator.go]: nit: `failed to create scope for machine %q` ?

File: pkg/apis/azureprovider/v1alpha1/azuremachineproviderconfig_types.go [hunk 10]
```diff
@@ -21,6 +21,13 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// Annotation constants
+const (
+	// ClusterIDLabel is the label that a machineset must have to identify the
+	// cluster to which it belongs.
+	ClusterIDLabel = "machine.openshift.io/cluster-api-cluster"
```
Comment (enxebre) [inline pkg/apis/azureprovider/v1alpha1/azuremachineproviderconfig_types.go]: why is this in this PR?

File: pkg/cloud/azure/actuators/machine/actuator.go [hunk 11]
```diff
@@ -70,18 +98,20 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
 		CoreClient: a.coreClient,
 	})
 	if err != nil {
-		return errors.Errorf("failed to create scope: %+v", err)
+		return a.handleMachineError(machine, apierrors.CreateMachine("failed to create scope: %v", err), createEventAction)
 	}
 	defer scope.Close()
 
-	err = NewReconciler(scope).Create(context.Background())
+	err = a.reconcilerBuilder(scope).Create(context.Background())
```
Comment (enxebre) [inline pkg/cloud/azure/actuators/machine/actuator.go]: why this change `ReconcilerBuilder` in this commit?
